sidekick_core/lib/src/sidekick_context.dart

```diff
@@ -0,0 +1,295 @@
+  /// command (and the nested commands) is executed. The process only knows
+  /// about a single sidekick package.
+  ///
+  /// When testing, multiple tests are executed in the same process.
+  /// This leads to shared memory. If the cache would be
+  /// enabled outside of [SidekickCommandRunner] executing a single command,
+  /// we would cache the values for the first test and reuse it in all others.
+  /// But each test uses a different temp directory, resulting in different
+  /// environment variables. Those can't be cached across tests.
+  static SidekickContextCache _cache = SidekickContextCache.noCache();
+
+  /// The sidekick package inside the [projectRoot]
+  static SidekickPackage get sidekickPackage {
+    return _cache.getOrCreate('sidekickPackage', _findSidekickPackage);
+  }
+
+  /// Returns the name of the CLI
+  static String get cliName {
+    return sidekickPackage.cliName;
+  }
+
+  static SidekickPackage _findSidekickPackage() {
+    // Strategy 1: Use the environment variable SIDEKICK_PACKAGE_HOME injected
+    // by entryPoint shell script
+    final envPackage = _findSidekickPackageFromEnv();
+    if (envPackage != null) {
+      return envPackage;
+    }
+
+    // Strategy 2: Script is within the sidekick package
+    final scriptPackage = _findSidekickPackageFromScript();
+    if (scriptPackage != null) {
+      return scriptPackage;
+    }
+
+    // Fallback strategy: searching all parent folders until we find a dart package root
+    final discovery = _discoverProject();
+    return discovery.sidekickPackage;
+  }
+
+  static SidekickPackage? _findSidekickPackageFromEnv() {
+    final injectedPackageHome = env[_envPackageHome];
+    if (injectedPackageHome != null && injectedPackageHome.isNotBlank) {
+      // When called via entryPoint, immediately return the package
+      return SidekickPackage.fromDirectory(Directory(injectedPackageHome))!;
+    }
+    return null;
+  }
+
+  static SidekickPackage? _findSidekickPackageFromScript() {
+    final script = File(DartScript.self.pathToScript);
+    final scriptPath = script.uri.path;
+
+    // When CLI is run with compiled entryPoint: <sidekick package>/build/cli.exe
+    if (scriptPath.endsWith('/build/cli.exe')) {
+      return SidekickPackage.fromDirectory(script.parent.parent)!;
+    }
+
+    // When CLI is run with `dart bin/main.dart`: <sidekick package>/bin/main.dart
+    if (scriptPath.endsWith('/bin/main.dart')) {
+      return SidekickPackage.fromDirectory(script.parent.parent)!;
+    }
+
+    // in `UpdateCommand` when the latest `update_sidekick_cli.dart` is written to <sidekick package>/build/update.dart to be executed
+    if (scriptPath.endsWith('/bin/update.dart')) {
+      return SidekickPackage.fromDirectory(script.parent.parent)!;
+    }
+    return null;
+  }
+
+  /// The directory where the [entryPoint] is located
+  ///
+  /// This directory is considered the project root.
+  static Directory get projectRoot {
```

Comment:
Is this even a good name?

Reply:
I had a quick chat with ChatCPT. Alternative names might be `origin`, `projectFolder`, `topLevelDirectory`, `baseDirectory`.
Any opinion @Giuspepe?